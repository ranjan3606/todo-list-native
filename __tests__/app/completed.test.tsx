import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
// Change from named import to default import
import CompletedScreen from '../../app/(tabs)/completed';

// Mock required modules
jest.mock('@expo/vector-icons', () => ({
  FontAwesome: 'FontAwesome',
}));

jest.mock('@/constants/Colors', () => ({
  Colors: {
    light: { 
      background: '#fff', 
      text: '#000', 
      tint: '#007AFF',
      lightText: '#666',
      cardBackground: '#f0f0f0'
    },
    dark: { 
      background: '#000', 
      text: '#fff', 
      tint: '#0A84FF',
      lightText: '#999',
      cardBackground: '#1c1c1e'
    },
    primary: '#007AFF',
  }
}));

jest.mock('@/services/theme', () => ({
  useActualColorScheme: () => 'light',
}));

jest.mock('@/hooks/useTodos', () => ({
  useTodos: () => ({
    completedTodos: [],
    isLoading: false,
    refresh: jest.fn(),
  }),
}));

jest.mock('@/components/TaskCard', () => {
  return {
    TaskCard: jest.fn().mockImplementation(props => {
      const mockReact = require('react');
      return mockReact.createElement(mockReact.Fragment, null, `Task: ${props.item.title}`);
    })
  };
});

jest.mock('@/components/TaskForm', () => ({
  TaskForm: jest.fn(() => null),
}));

jest.mock('@/i18n', () => ({
  useTranslation: () => ({
    t: (key) => {
      const translations = {
        'task.completedTasks': 'Completed Tasks',
        'common.loading': 'Loading',
        'task.noCompletedTasks': 'No completed tasks',
        'completedTask.editTitle': 'Edit Completed Task',
        'completedTask.editPrompt': 'Do you want to restore and edit this task?',
        'common.cancel': 'Cancel',
        'completedTask.restoreEdit': 'Restore & Edit',
      };
      return translations[key] || key;
    }
  })
}));

jest.mock('@/hooks/useStableRefresh', () => ({
  useStableRefresh: jest.fn(),
}));

jest.mock('@/components/PageLayout', () => {
  return {
    PageLayout: jest.fn().mockImplementation(props => {
      const mockReact = require('react');
      const { Text, View } = require('react-native');
      return mockReact.createElement(
        View, 
        null, 
        mockReact.createElement(Text, null, `HEADER: ${props.title}`),
        props.children
      );
    })
  };
});

jest.mock('react-native/Libraries/Alert/Alert', () => ({
  alert: jest.fn(),
}));

describe('CompletedScreen', () => {
  // Reset mocks before each test
  beforeEach(() => {
    jest.clearAllMocks();
  });

  // Helper to mock the Date object
  const mockDateImplementation = (mockISOString = '2023-01-01T12:00:00Z') => {
    const fixedDate = new Date(mockISOString);
    const originalDate = global.Date;
    global.Date = jest.fn(() => fixedDate) as any;
    global.Date.UTC = originalDate.UTC;
    global.Date.parse = originalDate.parse;
    global.Date.now = jest.fn(() => fixedDate.getTime());
    return () => {
      global.Date = originalDate;
    };
  };

  it('renders correctly with empty completed tasks', () => {
    const { getByText } = render(<CompletedScreen />);
    
    expect(getByText('HEADER: Completed Tasks')).toBeTruthy();
    expect(getByText('No completed tasks')).toBeTruthy();
  });
  
  it('displays the loading state correctly', () => {
    jest.spyOn(require('@/hooks/useTodos'), 'useTodos').mockReturnValue({
      completedTodos: [],
      isLoading: true,
      refresh: jest.fn(),
    });
    
    const { getByText } = render(<CompletedScreen />);
    expect(getByText('Loading')).toBeTruthy();
  });
  
  it('renders a list of completed tasks', () => {
    const restoreDate = mockDateImplementation();
    
    jest.spyOn(require('@/hooks/useTodos'), 'useTodos').mockReturnValue({
      completedTodos: [
        { id: '1', title: 'Task 1', completed: true, dueDate: new Date().toISOString() },
        { id: '2', title: 'Task 2', completed: true, dueDate: new Date().toISOString() },
      ],
      isLoading: false,
      refresh: jest.fn(),
    });
    
    // This should work but we won't be able to see the tasks due to the mock of TaskCard
    // The important part is that the component renders without errors
    const { toJSON } = render(<CompletedScreen />);
    expect(toJSON()).toBeTruthy();
    
    restoreDate();
  });
  
  it('matches snapshot', () => {
    const restoreDate = mockDateImplementation();
    
    jest.spyOn(require('@/hooks/useTodos'), 'useTodos').mockReturnValue({
      completedTodos: [
        { id: '1', title: 'Task 1', completed: true, dueDate: new Date().toISOString() },
        { id: '2', title: 'Task 2', completed: true, dueDate: new Date().toISOString() },
      ],
      isLoading: false,
      refresh: jest.fn(),
    });
    
    const { toJSON } = render(<CompletedScreen />);
    expect(toJSON()).toMatchSnapshot();
    
    restoreDate();
  });
});