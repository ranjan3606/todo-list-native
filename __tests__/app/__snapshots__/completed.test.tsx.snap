// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompletedScreen matches snapshot 1`] = `
<View>
  <Text>
    HEADER: Completed Tasks
  </Text>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    contentContainerStyle={
      {
        "flexGrow": 1,
        "paddingBottom": 20,
      }
    }
    data={
      [
        {
          "completed": true,
          "dueDate": "2023-01-01T12:00:00.000Z",
          "id": "1",
          "title": "Task 1",
        },
        {
          "completed": true,
          "dueDate": "2023-01-01T12:00:00.000Z",
          "id": "2",
          "title": "Task 2",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    getItemLayout={[Function]}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
    windowSize={10}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        style={null}
      >
        Task: Task 1
      </View>
      <View
        onFocusCapture={[Function]}
        style={null}
      >
        Task: Task 2
      </View>
    </View>
  </RCTScrollView>
</View>
`;
